### ISSUE TITLE
Is there also a discharge strategy for EV to minimise network congestion?

### ISSUE THREAD

**Issue by ChaelKruip** [quintel/etmoses#174]
I think we formulated the 'load management' strategy for buffering technologies (EV, HPs) as follows:

"Buffer (charge battery or fill tank) if the network has 'space' for it."

This prevents congestion due to high demand peaks (as it tends to flatten out the peaks over a larger time-span). 

On the other hand, congestion due to demand could also be decreased by **discharging** EV batteries when the demand is high and there is enough in the battery.

In short: Do we have a strategy for discharging EV to minimise congestion on the network? If not, what is the criterion for EV batteries to discharge?

**Comment by antw** [comment]
> Is there also a discharge strategy for EV to minimise network congestion?

No. Congestion currently [affects consumption only](https://github.com/quintel/etmoses/blob/e33506675204dbed287372cb9f8a431c3f144b63/app/models/network/technology_path.rb#L42). For this specific situation this is not really a problem, and a more advanced TechnologyPath class could get us most of the way there; although it would be an ugly hack. The proper solution is described [as part of this Trello card](https://trello.com/c/IGOyx3p5/21-5-make-the-calculator-subnet-aware).

> If not, what is the criterion for EV batteries to discharge?

EV will discharge if there is a deficit of production, i.e. there is more demand than there is supply. P2P batteries will be discharged first, then EVs.

**Comment by ChaelKruip** [comment]
Thanks for clarifying this @antw :smile:

**Comment by antw** [comment]
> Is there also a discharge strategy for EV to minimise network congestion?

So... it turns out that the answer to this is: "sort of, yes". EVs don't have any *explicitly coded* behaviour to minimise congestion by discharging, but they do act in a way which has the same effect. Provided that the EV and consumer both exist beneath a shared capacity constraint, the EV will discharge, reducing the likelihood of an exceedance.



Here's our example network:

![screen shot 2015-06-25 at 16 45 46](https://cloud.githubusercontent.com/assets/4383/8358702/ae9142cc-1b59-11e5-8630-e6da6a9fffc9.png)

[The producer](https://cloud.githubusercontent.com/assets/4383/8358922/09c894b4-1b5b-11e5-9eac-48ff15f856ee.png) outputs -2000, the EV has a capacity of 1000, and [the consumer](https://cloud.githubusercontent.com/assets/4383/8358938/1aa12ae4-1b5b-11e5-9179-288b2923ccf2.png) is turned off for 4 steps and is then turned on, consuming 3000. Here's what the load across the parent node looks like:

![screen shot 2015-06-25 at 16 47 32](https://cloud.githubusercontent.com/assets/4383/8358747/eb50cc96-1b59-11e5-9964-321db35dc71a.png)

There is considerable excess production in the first four time steps, which is returned to the grid. When the consumer is turned on in step 4, there is insufficient production, and 1000 must be taken from the grid. This exceeds the capacity of the "Shared constraint" node.

If we turn storage on, the EV will charge by 1000 in each of the first four time steps. This load is then discharged in the final four steps, eliminating the exceedance on the "Shared constraint" node:

![screen shot 2015-06-25 at 16 50 54](https://cloud.githubusercontent.com/assets/4383/8358850/a63419a0-1b5a-11e5-8f53-c9684ebc62a7.png)

--

Sadly, this example is quite carefully constructed. The calculator prefers not to discharge EVs unless they *really* need to. If another supplier comes online, the EV will not discharge and capacity will still be exceeded. I think the "subnet aware" calculator will be needed if we want to give priority to nearby production.

--

*Side note: I realise that the "Shared constraint" node should also consider a load below -500 (which represents a production of 500) to be an exceedance.*

**Comment by ChaelKruip** [comment]
Closing.
